### ISSUE TITLE
Q: is there an example, given a class, generate an interface for it?

### ISSUE THREAD

**Issue by mw66** [INRIA/spoon#4822]
Q: is there an example, given a class, generate an interface for it?

Thanks!

**Comment by mw66** [comment]
found it:

https://appdoc.app/artifact/fr.inria.gforge.spoon/spoon-core/5.5.0/spoon/reflect/declaration/CtType.html#getReference

**Comment by mw66** [comment]
Hi,

I've createInterface(), which call will write a generate CtInterface to disk file?

**Comment by mw66** [comment]
The new interface is created from Factory:

```
    CtInterface itfc = getFactory().createInterface();
```

Should the Factory keep a reference of it? or where I should put it into the "system" for output?

**Comment by SirYwell** [comment]
You can set the parent of the interface to a package. If the interface has no parent, it can't be discovered from traversing packages and types.

So while a type has a `simpleName` attribute, it has no `package` attribute but only a parent.

**Comment by mw66** [comment]
https://spoon.gforge.inria.fr/mvnsites/spoon-core/apidocs/spoon/reflect/declaration/CtElement.html#setParent(spoon.reflect.declaration.CtElement)

Manually sets the parent element of the current element.

I haven't tried yet, but from the document this is not obvious at all, esp the parameter type is CtElement.

While the meaning of setPackage() / getPackage() is far more clear.

**Comment by slarse** [comment]
> While the meaning of setPackage() / getPackage() is far more clear.

There are several problems with such methods. Starting with the simple case of a top-level type in a file, `setPackage()` would simply delegate to `setParent()`. That is slightly problematic in the sense that it's not necessarily obvious to a user that by changing the package, they change the parent of the type.

With nested types, it becomes significantly more complicated. Imagine a type like so:
```java
interface A {
    interface B {
    }

    interface C {
    }
}
```

and you call `setPackage()` on `B`. That would also require us to change the package of `A` and `C` to keep the model consistent. Essentially, calling `setPackage()` on an arbitrary type would require setting the same package for all enclosing types as well as all nested types, including those nested inside the enclosing type(s).

Based on this, I don't think a `setPackage()` utility method is a net gain for usability, and implementing it is also a bit inconvenient.

**Comment by mw66** [comment]
First, why didn't use `setParent()` to make the implementation easy? As you said, "setPackage() would simply delegate to setParent()"?

Second, if there are real difficulties preventing `setParent()` implementation, can we add to the doc and some examples, suggesting user use  `setParent()`, or use the other factory method as I did; or even remove the factory method with empty argument `getFactory().createInterface()`?

**Comment by slarse** [comment]
> First, why didn't use `setParent()` to make the implementation easy? As you said, "setPackage() would simply delegate to setParent()"?

Like I said, this only applies to top-level types. For nested types, you'd need to do set the parent of the top-level type.

> Second, if there are real difficulties preventing `setParent()` implementation.

It's not prohibitively difficult to implement, but like I mentioned it leads to potentially confusing and unexpected behavior where changing the package of one interface causes other interfaces to also be assigned to the new package.

> can we add to the doc and some examples, suggesting user use `setParent()`

Sure, improvements to the docs are welcome PRs. For runnable examples we have the https://github.com/SpoonLabs/spoon-examples repository (which admittedly has been a bit neglected in the past few years).

> or even remove the factory method with empty argument `getFactory().createInterface()`?

No, that is not an option. It would be a significantly breaking change

**Comment by slarse** [comment]
@MartinWitt @monperrus @SirYwell @I-Al-Istannen Any opinion on the idea of adding a `setPackage()` method, probably to the `CtType` interface? For the reasons I listed above I don't favor it, but you may feel differently.

**Comment by I-Al-Istannen** [comment]
I agree with you, the `setPackage` method could have effects on types I would definitely *not* expect at first.

As far as I can tell the problem is that @mw66 expects the package to be directly stored in the type itself. With this, a `setPackage` method makes a lot of sense and the omission is a bit weird. Instead, spoon builds up the package dynamically by traversing the current model tree structure to the nearest top level type. Therefore, `setPackage` now needs to transform the tree structure instead of just updating a property.

I am not sure if we want to document this somehow and *how* we would make this clear.  
Adding some javadoc to `getPackage()` along the lines of the following might help, but doesn't really feel ideal either:
> The package of a type is defined by the parent of the outer-most enclosing type. To set the package, you need to change the parent of this type to point to a different package.
